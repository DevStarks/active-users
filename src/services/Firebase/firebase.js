import app from 'firebase/app';
import 'firebase/database';

const config = {
  apiKey:      process.env[process.env.NODE_ENV === 'test' ? 'REACT_APP_TEST_API_KEY' : 'REACT_APP_API_KEY'],
  databaseURL: process.env[process.env.NODE_ENV === 'test' ? 'REACT_APP_TEST_DATABASE_URL' : 'REACT_APP_DATABASE_URL']
};

class Firebase {
  constructor() {
    app.initializeApp(config);

    this.database = app.database()
  }
}

export default Firebase;
