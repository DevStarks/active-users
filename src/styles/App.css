.App {
  text-align: center;
}

.App-container {
  background-color: #CFEBDF;
  min-height: 100vh;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  font-size: calc(10px + 2vmin);
  color: black;
}

.App-link {
  color: black;
}

.App-content {
  width: 40%;
}

.App-userCount {
  font-size: 1.5em;
}
