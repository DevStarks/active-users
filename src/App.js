import React from 'react';
import PropTypes from 'prop-types'
import './styles/App.css';
import { withFirebase } from './services/Firebase';
import generateUid from './utils/uid'
import Cookies from 'js-cookie'
import {
  updateActiveUserCount,
  subscribeToActiveUserCount
} from './actions/AppActions'

class App extends React.Component {
  constructor(props) {
    super(props)

    this.state = {
      loading: true,
      activeUserCount: null,
      uid: null
    };
  }

  async componentDidMount() {
    const db = this.props.firebase.database;
    await this.setCookie();

    updateActiveUserCount(db, this.state.uid);
    subscribeToActiveUserCount(db, this.onUserCountUpdated.bind(this));
  }

  setCookie() {
    if(!Cookies.get('uid')){
      Cookies.set('uid', generateUid());
    }
    this.setState({ uid: Cookies.get('uid') });
  }

  onUserCountUpdated(newCount) {
    this.setState({
      activeUserCount: newCount,
      loading: false
    });
  }

  renderUserCount() {
    return (
      <div className="App-content">
        <h1>Welcome!</h1>
        <h3>Including yours, this many browsers are currently accessing this app:</h3>
        <span className="App-userCount">{this.state.activeUserCount}</span>
      </div>
    );
  }

  renderLoading() {
    return (
      <p>Loading...</p>
    );
  }

  render() {
    return (
      <div className="App">
        <div className="App-container">
          {this.state.loading ? this.renderLoading() : this.renderUserCount()}
        </div>
      </div>
    );
  }
}

App.propTypes = {
  firebase: PropTypes.object.isRequired
};

export default withFirebase(App);
