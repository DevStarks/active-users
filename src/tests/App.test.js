import React from 'react';
import App from '../App';
import renderer from 'react-test-renderer';
import Firebase, { FirebaseContext } from '../services/Firebase'

it('renders', () => {
  const tree = renderer
    .create(
      <FirebaseContext.Provider value={new Firebase()}>
        <App />
      </FirebaseContext.Provider>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
