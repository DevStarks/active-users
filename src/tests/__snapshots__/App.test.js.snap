// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders 1`] = `
<div
  className="App"
>
  <div
    className="App-container"
  >
    <p>
      Loading...
    </p>
  </div>
</div>
`;
